Output Files
============
BACTpipe pipeline outputs several files in specific folders within the ``BACTpipe_results`` directory:

Output folders
**************

1. mash.screen
2. bbduk
3. fastqc
4. shovill
5. prokka
6. multiqc

Description of file outputs
***************************

Table below provides a description of the most relevant files in each above mentioned folder.

+-----------------+---------------------------+----------------------------------------------------------------------------------------------------------------+
| Output Folder   | File output               | File Description                                                                                               |
+=================+===========================+================================================================================================================+
| mash.screen     | - *.screening_results.tsv | - species likely present in sample, indicated by ``PASS`` or ``FAIL`` if pure or mixed isolate respectively    | 
+-----------------+---------------------------+----------------------------------------------------------------------------------------------------------------+
| bbduk           | - *.trimmed.fastq.gz files| - contains a pair of trimmed ``fastq.gz`` reads for each sample                                                |
|                 | - *.stats.txt             | - statistics relating to fraction of reads that matched each reference sequence                                |
|                 |                           |                                                                                                                |
+-----------------+---------------------------+----------------------------------------------------------------------------------------------------------------+ 
| fastqc          | - *.trimmed_fastqc.html   | - QC report for each sample showing how good or bad is your raw data                                           |
+-----------------+---------------------------+----------------------------------------------------------------------------------------------------------------+
| shovill         | - *.contigs.fa            | - most important output file showing the final corrected *de novo* assembly from each sample                   |
|                 | - *.assembly_stats.txt    | - summary of assembly statistics from each sample                                                              |
+-----------------+---------------------------+----------------------------------------------------------------------------------------------------------------+
| prokka          | - *.gff                   | - annotation in GFF3 format, containing both sequences and annotations                                         |
|                 | - *.gbk                   | - standard Genbank file							                                       |
|                 | - *.fna                   | - Nucleotide FASTA file of the input contig sequences						               |
|                 | - *.faa                   | - Protein FASTA file of the translated CDS sequences                             		               |
|                 | - *.ffn                   | - Nucleotide FASTA file of all the prediction transcripts (CDS, rRNA, tRNA, tmRNA, misc_RNA)                   |
|                 | - *.txt                   | - Statistics relating to the annotated features found                                                          |
+-----------------+---------------------------+----------------------------------------------------------------------------------------------------------------+
| multiqc         | - multiqc-report.html     | - summary statistics from results generated by prokka and fastqc tools                                         |
+-----------------+---------------------------+----------------------------------------------------------------------------------------------------------------+


**N.B** 
You will find ``BACTpipe_report.html`` within the main run folder after the pipeline is finished.
The report shows summary of overall execution time, distributions of resource usages and all executed tasks and their respective run time metrics. 


